### ISSUE TITLE
Unable to override bootstrap using className

### ISSUE THREAD

**Issue by Siegrift** [reactstrap/reactstrap#533]
## Unable to override bootsrap css properties when using className.

Inner implementation of reactstrap components composes user className with bootsrap's. However, the former is listed before the latter.

`  const classes = mapToCssModules(classNames(
    className,
    'navbar-brand'
  ), cssModule);`

This way it's impossible to change the bootstrap css as it overrides user defined css. 

 User className should be listed last.

**Comment by TheSharpieOne** [comment]
What [`classNames`](https://github.com/JedWatson/classnames) does it is not overriding, it is combining.
CSS is not code and the order does not matter in the class name. `.my-class.navbar-brand` is the same as`.navbar-brand.my-class`. The only time order matters is when you bring in your styles, but only when the [specificity value](https://css-tricks.com/specifics-on-css-specificity/#article-header-id-0) is the same.
So if you want to override the styles, you can change bring in bootstrap to your project and customize bootstrap's styles in SASS/SCSS. Or you can create another, more specific CSS selector (by combining your custom class name with bootstrap's class name) to have your styles override bootstrap's styles. Lastly, since many of reactstrap's components just simply add bootstrap's classes to components, if you do not want bootstrap's class, you can not use a reactstrap component. (There is also a trick with cssModules where you can map the bootstrap class name to a value you provide, but that is not the correct use).

**Example**
reactstrap
`<NavbarBrand>stuff</NavbarBrand>` -> `<a class="navbar-brand">stuff</a>`
If you don't want the `.navbar-brand` class you can
`<a>stuff</a>` -> `<a>stuff</a>`
(Note: output is the same, just without the `.navbar-brand` class)

**Example with extra className**
reactstrap
`<NavbarBrand className="my-class">stuff</NavbarBrand>` -> `<a class="my-class navbar-brand">stuff</a>`
If you don't want the `.navbar-brand` class you can
`<a className="my-class">stuff</a>` -> `<a class="my-class">stuff</a>`
(Note: output is the same, just without the `.navbar-brand` class)

tl;dr there is no way to remove/override the bootstrap classes names (since most of what reactstrap really does is add these class names). If you don't want the bootstrap class name, don't user a reactstrap component. If you want to override styles, that is CSS.

P.S. I am not saying do not use reactstrap at all, just that reactstrap doesn't make much since without the bootstrap classes; so much so that there is no way to remove them. For your use-case, reactstrap doesn't make much sense. And yes, there are components which do more than just add class names (such as `Modal`), but without bootstrap's class name and styles, the `Modal` isn't a modal. And really, without bootstrap's class names, it isn't bootstrap... and reactstrap is a bootstrap based component library.

**Comment by TheSharpieOne** [comment]
I'm closing this since the the issue is no longer active and it appear it was answered above. Please comment and it will be opened if you feel the issue has not been resolved / the question has not been answered

**Comment by dsuare1** [comment]
I ran across this same issue recently; I was using the <ButtonDropdown> component and noticed that it appended a `btn-secondary` class which overrode my class of `btn-primary` because of the way I had declared my CSS rules; since it's a legacy project, I didn't want to modify the order of the declarations and risk messing up styles in other parts of the app.

Here's how I handled it (assuming there are multiple instances of the dropdown, which was my case):

`componentDidMount() {
      const dropdownButtons = document.getElementsByClassName('dropdown-class-name-here');
      Object.keys(dropdownButtons).forEach(elem => dropdownToggle[elem].classList.remove('btn-secondary');
}
`

**Comment by TheSharpieOne** [comment]
@dsuare1, it sounds like you are using `<Button className="btn-primary">` instead of `<Button color="primary">`. You should use the color prop to specify the color/variant. If you don't define a `color`, it will default to `secondary` which sounds like your case. If you use `color="primary"` you would avoid needing that `componentDidMount` and manually manipulating the DOM which will be problematic if the button re-renders.

**Comment by TheSharpieOne** [comment]
@dsuare1 add the color prop to the `DropdownToggle`

```jsx
<DropdownToggle color="primary">
```

https://stackblitz.com/edit/reactstrap-v6-howgar?file=Example.js

**Comment by dsuare1** [comment]
@TheSharpieOne Thanks! That did the trick; I appreciate your replies.

I'm still fairly new to React development, so I wasn't unaware of the potential problems from using `componentDidMount`; I haven't found any super clear answer online, but it looks like that particular lifecycle method will *only fire once* after the node has mounted, meaning that it won't re-fire on re-renders. Is that correct?

Thanks again!

**Comment by TheSharpieOne** [comment]
That is correct. So when `DropdownToggle` re-rendered, it would add back the class which was removed and the `componentDidMount` would not run again to remove it again.
